### DISCUSSION TITLE
Java as a Subscriber crash in receive data with IDL Nested

### DISCUSSION THREAD

**Discussion by hendraprawira** [OpenDDS/OpenDDS#4660] [Q&A] [answered]
i have an IDL like this :

```
module NTB2Data {
    
    struct DPTData {
        double depth;
        string source;
        unsigned short status;
    };

    @topic
    struct Message {
        unsigned long long time;
        DPTData DPT;
    };
};
```

but after i test with my publisher (java), it come crash in java, i didnt find the error line.
but after i check, the subscriber find the mh.value.DPT.status is null, i check again the publisher is already fill the value:

the Subscriber Class Listener is

```
public synchronized void on_data_available(DataReader reader) {
        MessageDataReader mdr = MessageDataReaderHelper.narrow(reader);
        if (mdr == null) {
            logging.error(String.valueOf(mdr),"on_data_available: narrow failed.");
            return;
        }

        MessageHolder mh = new MessageHolder(new Message());
        SampleInfoHolder sih = new SampleInfoHolder(new SampleInfo());
        sih.value.source_timestamp = new Time_t();
        NTB2StatusModel ntb2Data = new NTB2StatusModel();
        System.out.println(mh.value.time);
        while (true) {
            mdr.take_next_sample(mh, sih);
            int status = mdr.take_next_sample(mh, sih);
            if (status == RETCODE_OK.value) {
                if (sih.value.valid_data) {
                    System.out.println("LOL "+mh.value.DPT);
                    ntb2Data.setAvailability(1);
                    try {
                        if(hasZeroValue(ntb2Data)){
                            //yellow
                            System.out.println("YELLOW ");
                            ntb2Data.setAvailability(3);
                        }else{
                            //green
                            System.out.println("GREEN ");
                            ntb2Data.setAvailability(1);
                        }
                    } catch (IllegalAccessException e) {
                        throw new RuntimeException(e);
                    }
                    ntb2Data.setCreatedTime(generateTimestamp());
                    SetResourceData.setResourceModel("ntb2", ntb2Data.getAvailability());
                    broadcastMessage("/ntb2-data", ntb2Data.toString());
                }
            } else if (status == RETCODE_NO_DATA.value) {
                return;
            } else {
                System.err.println("ERROR: take_next_sample_: " + status);
                return;
            }
        }
    }
```
anyone can help me? thanks
 forgive me for bad english, thanks

**Comment by mitza-oci** [comment]
https://github.com/OpenDDS/OpenDDS/blob/85561d54d10af95f6ff05b06e34a279983550600/java/FAQ#L68

**Reply by mitza-oci** [reply, marked as answer]
Some more specifics here:
`struct Message` in IDL turns in to `public class Message` in Java.
`Message` in IDL contains a member that's a `DPT` value.  It's not a "pointer to" or a "reference to" a DPT value.  
Java has no way to represent this data model, so it turns into a pointer which may be null.  Since IDL doesn't have a concept of null here, values of Message which contain a null pointer may not be passed into IDL-generated code like this `mdr.take_next_sample(mh, sih);`

**Reply by yuan-ge** [reply]
Dose this mean the custom data structure feature is not supported when used with Java?

**Reply by mitza-oci** [reply]
It does not mean that.

**Reply by yuan-ge** [reply]
So, what's the solution of the problem?🤔

**Reply by hendraprawira** [reply]
i'm already find the solution,
the solution is the struct of the message must be assign before calling take_next_sample(). like this:
` public synchronized void on_data_available(DataReader reader) {
        MessageDataReader mdr = MessageDataReaderHelper.narrow(reader);
        if (mdr == null) {
            logging.error(String.valueOf(mdr),"on_data_available: narrow failed.");
            return;
        }
        SampleInfoHolder sih = new SampleInfoHolder(new SampleInfo());
        sih.value.source_timestamp = new Time_t();
        NTB2StatusModel ntb2Data = new NTB2StatusModel();

        Message msg = new Message();

        msg.DPT = new DPTData();
        msg.HDT = new HDTData();
        msg.GLL = new GLLData();
        msg.MWV = new MWVData();
        msg.RMC = new RMCData();
        msg.VTG = new VTGData();
        msg.WIXDR = new WIXDRData();
        msg.ZDA = new ZDAData();
        msg.VHW = new VHWData();
        msg.PANZHRP = new PANZHRPData();
        MessageHolder mh = new MessageHolder(msg);

        while (true) {
            int status = mdr.take_next_sample(mh, sih);
            if (status == RETCODE_OK.value) {
                if (sih.value.valid_data) {
                    System.out.println(mh.value);
                }
            } else if (status == RETCODE_NO_DATA.value) {
                return;
            } else {
                System.err.println("ERROR: take_next_sample_: " + status);
                return;
            }
        }
    }`

**Reply by yuan-ge** [reply]
Oh, I know it, you have to init the class var when first use it. The constrcutor of generated class will not init the struct member.
